Libraries/Opc.Ua.Client/Session/Session.cs

```diff
@@ -2892,6 +2911,76 @@ public void ReadValues(
             }
         }
 
+        /// <inheritdoc/>
+        public byte[] ReadByteStringInChunks(NodeId nodeId)
+        {
+            if(ServerMaxByteStringLength <= 1)
+            {
+                throw new ServiceResultException(StatusCodes.BadIndexRangeNoData, "Server side MaxByteStringLength is not known or too small for reading data in chunks.");
+            }
+            int offset = 0;
+            List<byte[]> bytes = new List<byte[]>();
+
+            while(true)
+            {
+                ReadValueId valueToRead = new ReadValueId {
+                    NodeId = nodeId,
+                    AttributeId = Attributes.Value,
+                    IndexRange = new NumericRange(offset, offset + (int)ServerMaxByteStringLength - 1).ToString(),
+                    DataEncoding = null
+                };
+                ReadValueIdCollection readValueIds = new ReadValueIdCollection { valueToRead };
+
+                ResponseHeader responseHeader = Read(
+                    null,
+                    0,
+                    TimestampsToReturn.Neither,
+                    readValueIds,
+                    out DataValueCollection results,
+                    out DiagnosticInfoCollection diagnosticInfos);
+
+                ClientBase.ValidateResponse(results, readValueIds);
+                ClientBase.ValidateDiagnosticInfos(diagnosticInfos, readValueIds);
+
+                if(offset == 0)
+                {
+                    Variant wrappedValue = results[0].WrappedValue;
+                    if(wrappedValue.TypeInfo.BuiltInType != BuiltInType.ByteString ||
+                        wrappedValue.TypeInfo.ValueRank != ValueRanks.Scalar)
+                    {
+                        throw new ServiceResultException(StatusCodes.BadTypeMismatch, "Value is not a ByteString scalar.");
+                    }
+                }
+
+                if (StatusCode.IsBad(results[0].StatusCode))
+                {
+                    if (results[0].StatusCode == StatusCodes.BadIndexRangeNoData)
+                    {
+                        // this happens when the previous read has fetched all remaining data
+                        break;
+                    }
+                    ServiceResult serviceResult = ClientBase.GetResult(results[0].StatusCode, 0, diagnosticInfos, responseHeader);
+                    throw new ServiceResultException(serviceResult);
+                }
+
+                byte[] chunk = results[0].Value as byte[];
+                if(chunk.Length == 0)
```

Comment:
null check

Reply:
Will be taken care of with the next commit.